### DISCUSSION TITLE
What does _state.cpu_local_snapshot.particles.net_force store with the rigid constraint?

### DISCUSSION THREAD

**Discussion by XinyanYang95** [glotzerlab/hoomd-blue#1765] [General]
Hi All,

I'm running MD simulations with rigid bodies. Each rigid body contains several constituent particles. I want to fix the positions of some rigid particles, so I only added center particles of moving rigid bodies to the integrator. The simulation runs well. Fixed rigid particles are at their initial locations and other rigid particles are moving as expected. 

Then I want to collect forces on those fixed rigid particles. I checked  **_state.cpu_local_snapshot.particles.net_force**, there are forces on **constituent particles** and **center particles**. However, summing all forces on constituent particles does not equal to forces on center particles. So I'm confused what are the forces on those rigid bodies. Should I sum up forces on constituent particles as well as the center particle to get the total force on a rigid body?

Best,
Xinyan

**Comment by XinyanYang95** [comment]
BTW, I'm running GPU simulations with -np 1 (single CPU).

**Comment by XinyanYang95** [comment]
Does the net_force include forces from rigid constraint? If I sum up net_forces of constituent particles and center particles, the forces from rigid constraint will be canceled out?

**Comment by joaander** [comment]
`net_force` should be 0 on all constituent particles. `constrain.Rigid` adds the forces from all constituents to the net force on the central particle. It also zeros the net force on all constituents: https://github.com/glotzerlab/hoomd-blue/blob/abfaac977f30f6a880cbd824df0264d783d3f22e/hoomd/md/ForceComposite.cc#L748-L838
